core/base/inc/TObject.h

```diff
@@ -86,7 +86,16 @@ class TObject {
    enum {
       kSingleKey     = BIT(0),        ///< write collection with single key
       kOverwrite     = BIT(1),        ///< overwrite existing object with same name
-      kWriteDelete   = BIT(2)         ///< write object, then delete previous key with same name
+      kWriteDelete   = BIT(2),        ///< write object, then delete previous key with same name
+
+      ///< Used to request that the class specific implementation of `TObject::Write` 
+      ///< just prepare the objects to be ready to be written but do not actually write
+      ///< them into the TBuffer. This is just for example by TBufferMerger to request
+      ///< that the TTree inside the file calls `TTree::FlushBaskets` (outside of the merging lock)
+      ///< and TBufferMerger will later ask for the write (inside the merging lock).
+      ///< To take advantage of this feature the class needs to overload `TObject::Write`
+      ///< and use this enum value accordingly.  (See `TTree::Write` and `TObject::Write`)
+      kOnlyPrepStep  = BIT(3)         
```

Comment:
> For example, one possibility is that SetNotrees() doesn't match what's in the files being merged. 

SetNotrees is an actual (unlike what I mis-stated in one the comments) request to *ignore* TTrees and as such we can not automatically detect it but we also can not be "wrong" about it :).

> should go into some other place rather than in TObject

Technically it can't.  The using code (`hadd`, `TFileMerger`, etc) does not depend and does not know about TTree.
So the 2 alternatives I saw where:

* Reuse the interface and `TDirectory` traversal implemented by `TObject::Write` (and friends)
* Introduce yet another interface either in `TObject` or via `TClass` and have a 2nd implementation of the `TDirectory` traversal.

@Axel-Naumann and @amadio any opinion on the tradeoffs?

Reply:
It seems to me that `TFileMerger` does separate merging into several different cases for objects, directories, etc. Would it not be possible to add a new case for `TTree` there? Anyway, I leave this for you and Axel, as this touches low-level I/O things that you are better to judge than me.

Reply:
The issue is still the require indirection.  After review this with Axel, we concluded to migrate the functionality from `TObject::Write` to `*::Merge` (by adding an option there for `merge preparation`, it is still somewhat a stretch for the function but seems less disruptive (or more related to the core functionality) that to `TObject::Write`

Reply:
Yes, I think that moving this stuff to `::Merge` of each object makes a lot more sense as well.